components/brave_wallet/browser/eth_tx_manager.cc

```diff
@@ -906,28 +941,41 @@ void EthTxManager::SetNonceForUnapprovedTransaction(
 }
 
 std::unique_ptr<EthTxMeta> EthTxManager::GetTxForTesting(
+    const std::string& chain_id,
     const std::string& tx_meta_id) {
-  return GetEthTxStateManager()->GetEthTx(tx_meta_id);
+  return GetEthTxStateManager()->GetEthTx(chain_id, tx_meta_id);
 }
 
-void EthTxManager::OnNewBlock(uint256_t block_num) {
-  UpdatePendingTransactions();
+void EthTxManager::OnNewBlock(const std::string& chain_id,
+                              uint256_t block_num) {
+  UpdatePendingTransactions(chain_id);
 }
 
-void EthTxManager::UpdatePendingTransactions() {
+void EthTxManager::UpdatePendingTransactions(
+    const absl::optional<std::string>& chain_id) {
+  std::set<std::string> pending_chain_ids;
   size_t num_pending;
-  if (pending_tx_tracker_->UpdatePendingTransactions(&num_pending)) {
-    known_no_pending_tx_ = num_pending == 0;
-    CheckIfBlockTrackerShouldRun();
+  if (pending_tx_tracker_->UpdatePendingTransactions(chain_id, &num_pending,
+                                                     &pending_chain_ids)) {
+    if (num_pending == 0) {
+      known_no_pending_tx_ = true;
+      CheckIfBlockTrackerShouldRun(absl::nullopt);
+    } else {
+      known_no_pending_tx_ = false;
+      for (const auto& pending_chain_id : pending_chain_ids) {
+        CheckIfBlockTrackerShouldRun(pending_chain_id);
+      }
+    }
```

Comment:
It seems to me instead of `known_no_pending_tx_` we should keep `pending_chain_ids` a class field. Then update all trackers when this set changes or wallet lock state changes.

Reply:
That also allows us to do granular block tracker stop instead of stopping all only when there is no pending transactions.

Reply:
fixed in d16c41ba4c3b7a0c85ff99b9a5314e3c38a1d9ee